In [1]:
import pandas as pd 
import numpy as np 
df1=pd.read_csv('tmdb_5000_credits.csv')
df2=pd.read_csv('tmdb_5000_movies.csv')

### before i start the modelling i, am going to merge df1 and df2 on movie_id  and title, this is because these are the common columns in both DataFrames

In [2]:
print(df1.shape)
print(df1.info())
df1.head(2)

(4803, 4)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4803 non-null   int64 
 1   title     4803 non-null   object
 2   cast      4803 non-null   object
 3   crew      4803 non-null   object
dtypes: int64(1), object(3)
memory usage: 150.2+ KB
None


movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [3]:
print(df2.shape)
print(df2.info())
df2.head(2)

(4803, 20)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status    

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500

### to merge succesfully i simply had to rename 'movie_id' in df1 to 'id' so it can have same column name as df2

In [4]:
df1.columns = ['id','title','cast','crew']
df3= df2.merge(df1,on=['id','title'])
print(df3.shape)
print(df3.info())
df3.head(3)

(4803, 22)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 4803 entries, 0 to 4802
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status    

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  ...     revenue runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...  2787965087   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   961000000   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   880674609   148.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  

[3 rows x 22 columns]

### The modelling task 1. wants us to create a classifier that takes the crews job from a movie and tells u if the movie is an 'Animation' movie or 'Not_Animation" movie

To complete this task we can either train a Supervised or an Unsupervised machine learning Classifier. I would be trying the supervised machine learning approach here.
 
1. for the supervised machine learning approach, we need labels for the movies stating if they are Animation or Not_Animation. But from the Dataset such explicit labels are not given for a movies.

2. On a more closer Analysis of the dataset, you would notice that such labels can be extracted from the 'genres' column as there is a list containing all the genres of a given movie.

3. So for this i would define a function that iterates through that list and gives labels for each movie for if it is 'Animation' or 'Not_Animation'

4. For this task we would not be using all the features of the Dataset, we would be using only ('genres','original_title','crew') column, as these are were the attributes needed to ttrain the classifier are found


In [5]:
# selecting the features needed for the modeling and assigning to variable df
df = df3.loc[:,['genres','original_title','crew']]
print(df.info())
df.head(2)
# we can see there are no null values in the selected features

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4803 entries, 0 to 4802
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   genres          4803 non-null   object
 1   original_title  4803 non-null   object
 2   crew            4803 non-null   object
dtypes: object(3)
memory usage: 150.1+ KB
None


genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

#### Here i applied the literal_eval function to help with the iteration and extraction of data inside of the json columns 'genres' and 'crew'

In [6]:
from ast import literal_eval
col_names = ['genres','crew']
for cols in col_names:
    df[cols] = df[cols].apply(literal_eval)

#### This function here simply extracts all the genres from a movie and appends them to a list. stuffs like id and name are not needed here, we just need the genres.

In [7]:
def get_genres(x):
    if isinstance(x, list):
        names = [i['name'] for i in x]
        
        return names
df['genres'] = df['genres'].apply(get_genres)

#### For the crew members, we dont need their names or their sex, all we require is their job, and so the function below helps extract the jobs of all the crew members and appends them in a list. 

In [8]:
#function for crew memebers
def get_jobs(x):
    if isinstance(x, list):
        jobs = [i['job'] for i in x]
        
        return jobs

    #Return empty list in case of missing Data
df['crew'] = df['crew'].apply(get_jobs)

In [9]:
df.head(3)

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                                crew  
0  [Editor, Production Design, Sound Designer, Su...  
1  [Director of Photography, Director, Producer, ...  
2  [Original Music Composer, Director, Set Decora...

#### we could clearly see that the features have been extracted succesfully, but i must point out a few things.

#### 1. I would be using a Bag_of_Words model to count every individual job title in a row and record how many times they occur in that row 

#### 2. To Use the Bag of words model, The data has to be transformed to be in sentence form, where each row is a string containing

####        all the job titles as a single string seperated by commas or a white space, and not individual strings in a list.

#### 3. To achieve This i would define a function that takes text-strings in a list and combines them to a single string.

In [10]:
# column before the transformation
df['crew']

0       [Editor, Production Design, Sound Designer, Su...
1       [Director of Photography, Director, Producer, ...
2       [Original Music Composer, Director, Set Decora...
3       [Original Music Composer, Producer, Screenplay...
4       [Screenplay, Director, Thanks, Producer, Casti...
                              ...                        
4798    [Director, Director of Photography, Editor, Pr...
4799    [Director, Producer, Writer, Producer, Cinemat...
4800    [Costume Design, Producer, Director of Photogr...
4801                                   [Director, Writer]
4802    [Executive Producer, Executive Producer, Direc...
Name: crew, Length: 4803, dtype: object

In [11]:
def join_strings(x):
    return ",".join(x)
df['crew'] = df['crew'].apply(join_strings)

#### Notice the difference before and after the application of the function. The list have been removed.

In [12]:
df['crew']

0       Editor,Production Design,Sound Designer,Superv...
1       Director of Photography,Director,Producer,Scre...
2       Original Music Composer,Director,Set Decoratio...
3       Original Music Composer,Producer,Screenplay,Di...
4       Screenplay,Director,Thanks,Producer,Casting,So...
                              ...                        
4798    Director,Director of Photography,Editor,Produc...
4799    Director,Producer,Writer,Producer,Cinematograp...
4800    Costume Design,Producer,Director of Photograph...
4801                                      Director,Writer
4802    Executive Producer,Executive Producer,Director...
Name: crew, Length: 4803, dtype: object

#### Finally, this function helps to iterate through the genres column to check if a movie is animated or not.  

#### Note:

#### some of the movies genres might be written in lower case, so the function would first convert all the strings to lower case and 

#### then iterate through all the element to check for animation.


In [13]:
def find_animation(x):
    lst = []
    for i in x:
        i = i.lower()
        lst.append(i)
        
    for i in lst:
        if i == 'animation':
            return 'Animation'
    return 'Not_Animation'
df['labels'] = df['genres'].apply(find_animation)

In [14]:
# Transformed Dataset with the newly created labels feature
df.head(5)

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                crew         labels  
0  Editor,Production Design,Sound Designer,Superv...  Not_Animation  
1  Director of Photography,Director,Producer,Scre...  Not_Animation  
2  Original Music Composer,Director,Set Decoratio...  Not_Animation  
3  Original Music Composer,Producer,Screenplay,Di...  Not_Animation  
4  Screenplay,Director,Thanks,Producer,Casting,So...  Not_Animation

####  I would be using the following tranformers 
1. CountVectorizer also known as Bag_of_Words and would be including : ( stop_words, lemmatization and ngram_range) as hyperparameters.
2. I would include somme customized stopwords, and hence wont have need for tfidfTransformer for weight reduction
3. i will be using the MultinomialNaiveBayes model as the classifier ( its input features would be the output from the CountVectorizer).
4. Finally the whole process would be encapsulated into a scikit_learn pipeline
5. I would also split the Data and use some for training and the others for testing



In [15]:
print('The empty labels rows are: {}'.format(df.labels.isnull().sum()))
df.labels.value_counts()

The empty labels rows are: 0


Not_Animation    4569
Animation         234
Name: labels, dtype: int64

### I am going to try two approach

#### 1. Training the model on a slice of the observations because of tthe imbalance on the Animation and Not_Animation movies. This is solely for the usage of accuracy for evaluation of the performance of the classifier.

#### 2. The second approach i would fit the model to all the observations.

### Approach 1

#### So there are 234 movies with labels Animatin and 4569 with labels Not_animation, so i would take 134 of the observations for  Animation movies and 166 Not_animated movies to train the model. this way the classifier has an almost equal amount of documents to learn from. and we can also test the model later on the remmaining 100 observations of animated movie.

In [16]:
df1 = df[df.labels == 'Animation'].iloc[:134,:]
df2 = df[df.labels == 'Not_Animation'].iloc[:166,:]

In [17]:
print('The shape of df1 is: {}'.format(df1.shape))
print('\n')
print(f'The shape of df2 is:{df2.shape}')

The shape of df1 is: (134, 4)


The shape of df2 is:(166, 4)


#### So here i simply join the two dataframes and shuffle the dataset

In [18]:
from sklearn.utils import shuffle
df_slice = df1.append(df2)
df_slice = shuffle(df_slice)
df_slice = df_slice.reset_index(drop = True)
print(f'The shape of the Appended Slice is: {df_slice.shape}')

The shape of the Appended Slice is: (300, 4)


In [19]:
df_slice.head(2)

genres original_title  \
0  [Action, Adventure, Drama, Horror, Science Fic...         シン・ゴジラ   
1                        [Animation, Comedy, Family]     Free Birds   

                                                crew         labels  
0  Original Music Composer,Visual Effects,Art Dir...  Not_Animation  
1  Screenplay,Screenplay,Director,Screenplay,Writ...      Animation

#### lets take the remaining 100 observations of the 234 animated movies and the other non_animated observations and create a test df for testing after training.

In [20]:
test1 = df[df.labels ==  'Animation'].iloc[134:,:]
test2 = df[df.labels == 'Not_Animation'].iloc[166:,:]
print(f'The shape of test1 is:{test1.shape}')
print('The shape of test2 is: {}'.format(test2.shape))

The shape of test1 is:(100, 4)
The shape of test2 is: (4403, 4)


In [21]:
## Here i just join the two test_df( remember the model would not have seen this data)
test_df = pd.concat([test1,test2], axis= 0).reset_index(drop=True)
test_df.iloc[98:102,:]

genres               original_title  \
98              [Drama, Comedy, Animation]  I Married a Strange Person!   
99   [Animation, Family, Comedy, TV Movie]    A Charlie Brown Christmas   
100                     [Action, Thriller]            The Bourne Legacy   
101               [Action, Crime, Fantasy]               Batman & Robin   

                                                  crew         labels  
98                                    Director,Creator      Animation  
99                                     Director,Writer      Animation  
100  Producer,Original Music Composer,Casting,Execu...  Not_Animation  
101  Sound Designer,Casting,Author,Visual Effects S...  Not_Animation

In [22]:
test_df['crew'][41]

'Casting,Writer,Director of Photography,Original Music Composer,Editor,Costume Design,Producer,Producer,Director,Production Design,Executive Producer,Set Decoration,Associate Producer,Executive Producer,Associate Producer,Assistant Director'

#### Before i proceed i must define my documents and my labels. The document here means what we consider as a unit of analysis, i.e an observation, and i already pointed out that the job title of each row would be our document and the labels would be its accompanying label.That is we would have a total of 300 documents and labels. Now all that is left is to transform them into a form the BOW model would recognize which is an array.

In [23]:
doc = df_slice['crew'].values
label = df_slice['labels'].values

In [24]:
import spacy
nlp = spacy.load('en_core_web_sm')

#### Stop words are words that are not to be counted in the bag of words model. spacy already has a stop_word estimator that is made of a dictionary of words, but it also allows us to add our own stop words to the dictionary.

#### I would be adding some very common words that appear in both documents refering to Animated movies and Non_Animated movies.

#### words like Editor, Director, Art appear multiple times in bothe documents and if they are counted would have very highe weights and hence make it dificult for the classifier to dintinguish between key elements.

#### Lemmatization simply implies Reducing Each words to its lemar. words like ( run, ran, runs, and running) would all be counted as run.

In [25]:
from spacy.lang.en import STOP_WORDS
stop_words_str = " ".join(STOP_WORDS)
stop_words_lemma = set(word.lemma_ for word in nlp(stop_words_str))

additional_words = ['Editor','Art','Sound','Producer','wigmaker','vfx',
             'Director','Acting','Crew','Camera','Writing','armorer','underwater',
             'Sound','lighting','Executive Producer','frame','writing','writer',
             'Production Design','Thanks','Casting','3d','administration',
             'Original Music Composer','Screenplay','24','administrator','aerial',
             'Additional Soundtrack','Associate Producer','&',',','-','/',
             'Set Designer','Set Production','Driver','Assistant','unit','adr',
             'utility','assistant','assist','transportation','timer','additional',
             'account','book','boom','system','teacher','technician','technical',
             'thank','editor','art','sound','producer','crew','camera','art','writing','acting',
             'boy','buyer','captain','capture','studio','stunt','stylist','boy','author',
             'video','title','supervise','supervisor','cg','stand','standby']

for word in additional_words:
    stop_words_lemma = stop_words_lemma.union({word})

In [26]:
def lemmatizer(text):
     return [word.lemma_ for word in nlp(text)]

In [61]:
from sklearn.feature_extraction.text import TfidfVectorizer,CountVectorizer,TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline





bow = CountVectorizer(ngram_range = (1,1), stop_words=stop_words_lemma)

tfidf = TfidfTransformer()


pipe = Pipeline([('bag_of_words', bow),('classifier', MultinomialNB())])
pipe.fit(doc,label)


print('Training accuracy For the Sliced DataFrame is: {}'.format(pipe.score(doc,label)))

C:\Users\Anthony\anaconda3\lib\site-packages\sklearn\feature_extraction\text.py:386: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['associate', 'casting', 'composer', 'design', 'designer', 'director', 'driver', 'executive', 'music', 'original', 'production', 'pron', 'screenplay', 'set', 'soundtrack', 'thanks'] not in stop_words.
  'stop_words.' % sorted(inconsistent))


Training accuracy For the Sliced DataFrame is: 0.977930460129086


#### The test accuracy is quite good, and since there is not much inbalance in the dataset the metric is acceptable. Now lets train the model using all the observations.

#### Note!! 
#### you can go to the bottom of the notebook to test the model usingg the test df we defined above.

#### The cell below is just an illustration of what the countvectorizer does.

#### It basically takes each words and counts its frequency of occurence in a document, each row in the Dataframe below represents a movie and all the job role occurences for that movie, The words included in the stopwords are not included in the words count. For this illustration i included a tfidf transformer.

In [39]:
bow = CountVectorizer(ngram_range = (1,1),min_df = 5, stop_words=stop_words_lemma, tokenizer = lemmatizer, max_features = 100)
X = bow.fit_transform(doc)
tfidf = TfidfTransformer()
F = tfidf.fit_transform(X)
classifier =  MultinomialNB()
model = classifier.fit(F,label)

C:\Users\Anthony\anaconda3\lib\site-packages\sklearn\feature_extraction\text.py:386: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['-pron-', 'act', 'associate', 'cast', 'composer', 'design', 'designer', 'director', 'driver', 'executive', 'music', 'original', 'production', 'screenplay', 'set', 'soundtrack', 'thanks'] not in stop_words.
  'stop_words.' % sorted(inconsistent))


In [40]:
values = X.toarray()
values.shape

(300, 100)

In [41]:
## showing some of the counted words
headers = bow.get_feature_names()
headers[:10]

['accountant',
 'animation',
 'animator',
 'artist',
 'associate',
 'carpenter',
 'cast',
 'casting',
 'character',
 'choreographer']

In [42]:
dataframe =  pd.DataFrame(values, columns = headers)
dataframe

accountant  animation  animator  artist  associate  carpenter  cast  \
0             0          0         0       0          0          0     0   
1             0          6         0       1          0          0     0   
2             0          0         0       0          4          0     5   
3             0         11         1       1          0          0     0   
4             0          1         0       0          0          0     0   
..          ...        ...       ...     ...        ...        ...   ...   
295           0          0         0       0          0          0     0   
296           0          0         0       0          0          0     0   
297           0          1         0       8          0          0     2   
298           0          0         1       2          0          0     2   
299           0         12         0       3          0          0     0   

     casting  character  choreographer  ...  second  sequence  service  set  \
0          1          0              0  ...       0         0        0    2   
1          0          0              0  ...       0         0        0    0   
2          1          1              0  ...       1         0        0    6   
3          0          0              0  ...       0         0        0    0   
4          1          0              1  ...       0         0        0    0   
..       ...        ...            ...  ...     ...       ...      ...  ...   
295        0          0              0  ...       0         0        0    0   
296        0          0              0  ...       0         0        0    0   
297        0          0              1  ...       1         0        0    7   
298        1          0              1  ...       0         0        0   11   
299        2          0              0  ...       0         1        0    0   

     software  special  steadicam  story  storyboard  visual  
0           0        1          0      0           0       5  
1           0        0          0      0           1       0  
2           0        1          1      0           0      15  
3           0        0          0      0           0       0  
4           0        0          0      0           0       0  
..        ...      ...        ...    ...         ...     ...  
295         0        0          0      0           0       0  
296         0        0          0      0           0       0  
297         0        3          1      0           0      15  
298         0        0          0      0           0      14  
299         0        1          0      3           0       1  

[300 rows x 100 columns]

### Approach 2

In [43]:
# Transformed Dataset with the newly created labels feature
df.tail(5)

genres             original_title  \
4798           [Action, Crime, Thriller]                El Mariachi   
4799                   [Comedy, Romance]                  Newlyweds   
4800  [Comedy, Drama, Romance, TV Movie]  Signed, Sealed, Delivered   
4801                                  []           Shanghai Calling   
4802                       [Documentary]          My Date with Drew   

                                                   crew         labels  
4798  Director,Director of Photography,Editor,Produc...  Not_Animation  
4799  Director,Producer,Writer,Producer,Cinematograp...  Not_Animation  
4800  Costume Design,Producer,Director of Photograph...  Not_Animation  
4801                                    Director,Writer  Not_Animation  
4802  Executive Producer,Executive Producer,Director...  Not_Animation

#### Let's define our labels and document, we can see that some labels are empty, so we would fill such rows with an  empty string

In [44]:
doc = df['crew'].values
# lets show the document(job  titles) for NewlyWeds index number 4799
doc[4799]

'Director,Producer,Writer,Producer,Cinematography,Music,Editor,Producer'

In [45]:
label = df['labels'].values


In [54]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(doc,label,test_size = 0.2, random_state = 10)

In [55]:
(pd.Series(y_test)).value_counts()

Not_Animation    914
Animation         47
dtype: int64

In [56]:
from sklearn.feature_extraction.text import TfidfVectorizer,CountVectorizer,TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline





bow = CountVectorizer(ngram_range = (1,1), stop_words=stop_words_lemma)




pipe = Pipeline([('bag_of_words', bow),('classifier', MultinomialNB())])
pipe.fit(x_train,y_train)


print('Training accuracy: {}'.format(pipe.score(x_train,y_train)))

C:\Users\Anthony\anaconda3\lib\site-packages\sklearn\feature_extraction\text.py:386: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['associate', 'casting', 'composer', 'design', 'designer', 'director', 'driver', 'executive', 'music', 'original', 'production', 'pron', 'screenplay', 'set', 'soundtrack', 'thanks'] not in stop_words.
  'stop_words.' % sorted(inconsistent))


Training accuracy: 0.9770952628839146


In [57]:
y_pred = pipe.predict(x_test)

#### The test accuracy is improved here, although there is an imbalance but our initial test validates this.

In [58]:
from sklearn.metrics import accuracy_score, recall_score
accuracy = accuracy_score(y_test, y_pred)
print('The Accuracy on The Test Set is: %s' % accuracy)

The Accuracy on The Test Set is: 0.9791883454734651


In [62]:
from sklearn.metrics import classification_report, confusion_matrix
report = classification_report(y_test,y_pred)
print(report)

               precision    recall  f1-score   support

    Animation       0.83      0.72      0.77        47
Not_Animation       0.99      0.99      0.99       914

     accuracy                           0.98       961
    macro avg       0.91      0.86      0.88       961
 weighted avg       0.98      0.98      0.98       961



In [63]:
matrix = confusion_matrix(y_test, y_pred)
print(matrix)

[[ 34  13]
 [  7 907]]


#### You can use this to check for a particular Movie Title just change the row index number for a different movie 

In [277]:
test = test_df.drop(columns='labels')

In [279]:
movie_name = test.loc[30,'original_title']
movie_name

'Sausage Party'

#### You can use this to get the crew Job.

In [281]:
crew_job = test.loc[30,'crew']
crew_job

'Director,Producer,Original Music Composer,Executive Producer,Story,Story,Producer,Screenplay,Original Music Composer,Story,Producer,Screenplay,Director,Producer,Screenplay,Executive Producer,Screenplay,Executive Producer,Layout,Line Producer,Co-Producer,Co-Producer,Co-Producer,Storyboard,Storyboard,Executive Producer,Editor,Art Direction,Production Manager,Animation,VFX Artist,Lighting Technician,Layout,Animation,Animation Supervisor,Animation Manager,Administration,Animation,Animation,Animation,Animation,Animation,Animation,Animation,Music Supervisor,Visual Effects Supervisor'

#### And finally you can check if a movie is animation or not by Passing the crew's_job as an argument in a list here

In [282]:
y_pred = pipe.predict([crew_job])
y_pred

array(['Animation'], dtype='<U13')

### you can also search for a random movie online, pass the jobs of the crew as an array delimited by space into the predictor, the model would tell you if that movie is an animation movie or not even without seeing the genre. all it needs to make predictions are the crew's job.